#### Name: Andrew Shapiro
#### Date: 12/04/2024
#### Exercise: Final Project, Milestone 3
#### Description: This notebook will read in a Board Games Dataset and clean it up so that we can analyze four research questions for our final project.

In [43]:
import pandas as pd

## 1. Using a Jupyter Notebook, import the data into a pandas dataframe using the appropriate Python library for the dataset file type.
For this assignment, I have a huge interest in board games, so that's what my final project is about.
* My dataset is in a file called `BGG.csv`. The link to the original website for the dataset is https://ieee-dataport.org/open-access/boardgamegeek-dataset-board-games.

In [44]:
# Read the dataset into the pandas data frame.
df_csv = pd.read_csv('BGG.csv')

## 2. Perform data cleaning operations on the dataset.
These are the provided instructions for this portion of the assignment.
* Did your exploratory analysis from last week highlight anything that needs to be cleaned?
* Investigate any NaN's that you find. Should these be replaced with another value? If so, what would be the best value? Should that record be deleted instead?
* Are there outliers to any of the data items? What could explain those outliers? Should those records be changed or deleted?
* Are the column names not in a format that makes it easy to refer to?
* Do you need to merge multiple datasets into one?
* Do you need to split up a column's data into multiple columns or simplify its values?

### **(a)** Did your exploratory analysis from last week highlight anything that needs to be cleaned?
There are a few things that my exploratory analysis from last week highlighted.

1. It highlighted that the last two columns of my dataset ('Mechanics' and 'Domains') were not tidy. I will clean these columns when answering the final question **(f)** about splitting a column's data into multiple columns or simplify its values.
2. It showed me that there were a few "N/A"s that I will look into and clean up accordingly in the next question. **(b)**
3. There was a few outliners that went into BC (negative year value) which I believe is not ideal which I will clean first with my analysis (see below).

I will be fixing all three of these later on in this notebook.

First of all, before we get into more deep analysis, I noticed in my Milestone #2 that my dataframe was a little bit big. After looking more into the dataset, I noticed that the range of years was huge and all of it is not necessary to board game developers today who want to release a board game that could be popular in modern day. For this reason, I have decided to limit my scope to only include board games that were created **in the past 50 years** so that we can get accurate, but results that are more reflective now rather than all of history.

In [45]:
df = df_csv[df_csv['Year Published'] >= 1974]

df

ID                                   Name  Year Published  \
0      174430.0                             Gloomhaven          2017.0   
1      161936.0              Pandemic Legacy: Season 1          2015.0   
2      224517.0                      Brass: Birmingham          2018.0   
3      167791.0                      Terraforming Mars          2016.0   
4      233078.0      Twilight Imperium: Fourth Edition          2017.0   
...         ...                                    ...             ...   
20316    5050.0  The Worst-Case Scenario Survival Game          2001.0   
20325  205322.0             The Oregon Trail Card Game          2016.0   
20326    3510.0                    Battle of the Sexes          1997.0   
20328    5895.0                   Hungry Hungry Hippos          1978.0   
20334    3522.0                                    LCR          1983.0   

       Min Players  Max Players  Play Time  Min Age  Users Rated  \
0                1            4        120       14        42055   
1                2            4         60       13        41643   
2                2            4        120       14        19217   
3                1            5        120       12        64864   
4                3            6        480       14        13468   
...            ...          ...        ...      ...          ...   
20316            2           10         40        8          588   
20325            2            6         45       12         2654   
20326            2            8         45       12         1090   
20328            2            4         10        4         2361   
20334            3           12         20        5         1835   

       Rating Average  BGG Rank  Complexity Average  Owned Users  \
0                8.79         1                3.86      68323.0   
1                8.61         2                2.84      65294.0   
2                8.66         3                3.91      28785.0   
3                8.43         4                3.24      87099.0   
4                8.70         5                4.22      16831.0   
...               ...       ...                 ...          ...   
20316            3.32     20318                1.20       1148.0   
20325            4.27     20327                1.18       7183.0   
20326            3.48     20328                1.08       1987.0   
20328            4.12     20330                1.05       2568.0   
20334            3.42     20336                1.05       3441.0   

                                               Mechanics  \
0      Action Queue, Action Retrieval, Campaign / Bat...   
1      Action Points, Cooperative Game, Hand Manageme...   
2      Hand Management, Income, Loans, Market, Networ...   
3      Card Drafting, Drafting, End Game Bonuses, Han...   
4      Action Drafting, Area Majority / Influence, Ar...   
...                                                  ...   
20316                                       Dice Rolling   
20325  Communication Limits, Connections, Cooperative...   
20326                                    Team-Based Game   
20328                                                NaN   
20334                                       Dice Rolling   

                              Domains  
0      Strategy Games, Thematic Games  
1      Strategy Games, Thematic Games  
2                      Strategy Games  
3                      Strategy Games  
4      Strategy Games, Thematic Games  
...                               ...  
20316                     Party Games  
20325                    Family Games  
20326                     Party Games  
20328                Children's Games  
20334   Children's Games, Party Games  

[19494 rows x 14 columns]

With this, our scope is now smaller and the data is now more relevant to modern day board game developers. Additionally, our outliers of BC data is now gone with this change as well. Hopefully, this will make our data cleaning and analysis not only easier to conduct, but also provide modern results for board game developers.

Additionally, a few columns seem redundant for my analysis. These columns are the 'ID' (which corresponds to the website, which, in my analysis, is not needed) and the 'BGG Rank' (which is also related to the dataset's website, BGG, that will not be of use for our analysis). I will now go ahead and drop these columns from my dataset.

In [46]:
# Drop the 'ID' and 'BGG Rank' columns from our dataset.
cleaning_df = df.drop(columns=['ID', 'BGG Rank'])

cleaning_df

Name  Year Published  Min Players  \
0                                 Gloomhaven          2017.0            1   
1                  Pandemic Legacy: Season 1          2015.0            2   
2                          Brass: Birmingham          2018.0            2   
3                          Terraforming Mars          2016.0            1   
4          Twilight Imperium: Fourth Edition          2017.0            3   
...                                      ...             ...          ...   
20316  The Worst-Case Scenario Survival Game          2001.0            2   
20325             The Oregon Trail Card Game          2016.0            2   
20326                    Battle of the Sexes          1997.0            2   
20328                   Hungry Hungry Hippos          1978.0            2   
20334                                    LCR          1983.0            3   

       Max Players  Play Time  Min Age  Users Rated  Rating Average  \
0                4        120       14        42055            8.79   
1                4         60       13        41643            8.61   
2                4        120       14        19217            8.66   
3                5        120       12        64864            8.43   
4                6        480       14        13468            8.70   
...            ...        ...      ...          ...             ...   
20316           10         40        8          588            3.32   
20325            6         45       12         2654            4.27   
20326            8         45       12         1090            3.48   
20328            4         10        4         2361            4.12   
20334           12         20        5         1835            3.42   

       Complexity Average  Owned Users  \
0                    3.86      68323.0   
1                    2.84      65294.0   
2                    3.91      28785.0   
3                    3.24      87099.0   
4                    4.22      16831.0   
...                   ...          ...   
20316                1.20       1148.0   
20325                1.18       7183.0   
20326                1.08       1987.0   
20328                1.05       2568.0   
20334                1.05       3441.0   

                                               Mechanics  \
0      Action Queue, Action Retrieval, Campaign / Bat...   
1      Action Points, Cooperative Game, Hand Manageme...   
2      Hand Management, Income, Loans, Market, Networ...   
3      Card Drafting, Drafting, End Game Bonuses, Han...   
4      Action Drafting, Area Majority / Influence, Ar...   
...                                                  ...   
20316                                       Dice Rolling   
20325  Communication Limits, Connections, Cooperative...   
20326                                    Team-Based Game   
20328                                                NaN   
20334                                       Dice Rolling   

                              Domains  
0      Strategy Games, Thematic Games  
1      Strategy Games, Thematic Games  
2                      Strategy Games  
3                      Strategy Games  
4      Strategy Games, Thematic Games  
...                               ...  
20316                     Party Games  
20325                    Family Games  
20326                     Party Games  
20328                Children's Games  
20334   Children's Games, Party Games  

[19494 rows x 12 columns]

### **(b)** Investigate any NaN's that you find. Should these be replaced with another value? If so, what would be the best value? Should that record be deleted instead?
Since my dataset is large, in order to find how many rows have at least 1 row of data that has a NaN, I will be running a command. This should help me see if I missed any NaNs and which columns have NaNs that I will need to resolve. I did this by checking the .isnull() command and then allow it to return any column with at least 1 NaN.
  

In [47]:
# Check which rows have at least 1 missing value.
cleaning_df[cleaning_df.isnull().any(axis=1)]

Name  Year Published  Min Players  \
663               King of Tokyo: Dark Edition          2020.0            2   
1011               Smash Up: It's Your Fault!          2016.0            2   
1047    Exit: The Game - The Sinister Mansion          2018.0            1   
1059                     Timeline: Inventions          2010.0            2   
1073        Unmatched: Robin Hood vs. Bigfoot          2019.0            2   
...                                       ...             ...          ...   
20291                    Would You Rather...?          1998.0            3   
20295  Dirty Minds: The Game of Naughty Clues          1990.0            2   
20308                                 Hangman          1976.0            2   
20314                              Guess Who?          1979.0            2   
20328                    Hungry Hungry Hippos          1978.0            2   

       Max Players  Play Time  Min Age  Users Rated  Rating Average  \
663              6         30        8         1854            7.89   
1011             2         60       14         1663            7.53   
1047             4         90       12         1595            7.45   
1059             8         15        8         7257            6.71   
1073             2         20        9          924            8.24   
...            ...        ...      ...          ...             ...   
20291            8         20       12          678            4.37   
20295            6         45       18          563            4.04   
20308            2         10        6          942            4.21   
20314            2         20        6         4715            4.80   
20328            4         10        4         2361            4.12   

       Complexity Average  Owned Users  \
663                  1.63       5157.0   
1011                 2.06       6347.0   
1047                 2.61       3446.0   
1059                 1.11      12448.0   
1073                 1.93       2665.0   
...                   ...          ...   
20291                1.21       1823.0   
20295                1.15       1504.0   
20308                1.11       1187.0   
20314                1.12       7576.0   
20328                1.05       2568.0   

                                               Mechanics           Domains  
663    Card Drafting, Dice Rolling, King of the Hill,...               NaN  
1011   Area Majority / Influence, Hand Management, Ta...               NaN  
1047                                    Cooperative Game               NaN  
1059                                                 NaN      Family Games  
1073   Action Points, Hand Management, Line of Sight,...               NaN  
...                                                  ...               ...  
20291                                             Voting               NaN  
20295                                                NaN       Party Games  
20308                                                NaN      Family Games  
20314                                                NaN  Children's Games  
20328                                                NaN  Children's Games  

[10227 rows x 12 columns]

So far, I see that only the 'Mechanics' and 'Domains' columns of the dataset have 'NaN's. Since a few of my research questions do not focus on these columns, I have decided to keep the row (since it could be used for other data questions) and instead just rename all of the 'NaN's in these columns to be 'Unknown'. This will allow me to continue using the other data for other questions and if I approach a question that requires me to use either of these columns, I can simply filter out the data that's 'Unknown'.

In [48]:
# Remove any 'NaN's in the 'Mechanics' and 'Domains' columns and replace them with 'Unknown'
cleaning_df['Mechanics'] = cleaning_df['Mechanics'].fillna('Unknown')
cleaning_df['Domains'] = cleaning_df['Domains'].fillna('Unknown')

As a final test, I will check if there are still any 'NaN's that may not have made it in the preview.

In [49]:
# Check which rows have at least 1 missing value.
cleaning_df[cleaning_df.isnull().any(axis=1)]

Name  Year Published  \
2828                       Guildhall Fantasy: Fellowship          2016.0   
3590                         Guildhall Fantasy: Alliance          2016.0   
3739                        Guildhall Fantasy: Coalition          2016.0   
5807                                       Chariot Lords          1999.0   
9202         Operation Market Garden: Descent into Hell           1985.0   
9317                                                Hoax          1981.0   
10075              Devil Boats: PT Boats in the Solomons          2021.0   
10776                            Ace of Aces: Jet Eagles          1990.0   
10835                                 Die Erben von Hoax          1999.0   
11152  Rommel in North Africa: The War in the Desert ...          1986.0   
11669                  Migration: A Story of Generations          2012.0   
12649                   Die Insel der steinernen Wachter          2009.0   
12764                   Dragon Ball Z TCG (2014 edition)          2014.0   
13282                                           Dwarfest          2014.0   
14053                                        Contrario 2          2006.0   
14663                           Warage: Extended Edition          2017.0   
16292                                            Rainbow          2010.0   
17009                Sexy, el juego del arte del flirteo          2014.0   
18672                    Dracarys Dice Don't Get Burned!          2018.0   
19332  Battleship: Tactical Capital Ship Combat 1925-...          1986.0   
19474                          The Umbrella Academy Game          2020.0   
20040                                    Hidden Conflict          2005.0   

       Min Players  Max Players  Play Time  Min Age  Users Rated  \
2828             2            4         45       10          565   
3590             2            4         45       10          360   
3739             2            4         45       10          336   
5807             3            4        360       12          221   
9202             2            2        120       12           94   
9317             3           12         45       10          216   
10075            1            1         60       14           49   
10776            2            2         20       10          110   
10835            3            8         45       12          137   
11152            2            2          0       12           53   
11669            2            4         30       12           49   
12649            2            4        120       12           49   
12764            2            2         20        8           33   
13282            2            6         45       12           82   
14053            2           12          0       14           37   
14663            2            6         90       10           49   
16292            2            6         20        8           35   
17009            3            4        120       18           39   
18672            2            5         15       10           86   
19332            2            2         90        8          131   
19474            1            6         40       12           41   
20040            2            6         30       10          128   

       Rating Average  Complexity Average  Owned Users  \
2828             7.13                2.00          NaN   
3590             7.20                2.14          NaN   
3739             7.19                2.13          NaN   
5807             6.68                3.00          NaN   
9202             6.72                3.00          NaN   
9317             5.97                1.38          NaN   
10075            7.84                2.83          NaN   
10776            6.26                2.00          NaN   
10835            6.05                2.00          NaN   
11152            6.76                4.00          NaN   
11669            7.20                2.00          NaN   
12649            6.73                3.00 

Now, I can clearly see that the "Owned Users" column is also another column that has 'NaN's (while it doesn't have many, it still has some). Since most of my research also doesn't need the total users column, I have decided to keep the row (since it could be used for other data questions) and instead just rename all of the 'NaN's in these columns to be 'Unknown'. This will allow me to continue using the other data for other questions and if I approach a question that requires me to use the 'Total Users' column, I can simply filter out the data that's 'Unknown'.

In [50]:
# Remove any 'NaN's in the 'Owned' columns and replace them with 'Unknown'
cleaning_df['Owned Users'] = cleaning_df['Owned Users'].fillna('Unknown')

As a final test, I will check if there are still any 'NaN's that may not have made it in the preview.

In [51]:
# Check which rows have at least 1 missing value.
cleaning_df[cleaning_df.isnull().any(axis=1)]

Empty DataFrame
Columns: [Name, Year Published, Min Players, Max Players, Play Time, Min Age, Users Rated, Rating Average, Complexity Average, Owned Users, Mechanics, Domains]
Index: []

We have now cleaned up all of the 'NaN's in our dataset. As I stated earlier, I did not drop any value as it was not very important (not all research questions needed the column data) and since it had other data still available, I would still be able to reference the data it still has rather than drop the entire row.

Renaming all of our 'NaN's to 'Unknown' is a much better choice as it would allow me to continue using the other data for other questions and if I approach a question that requires me to use the 'Total Users' column, I can simply filter out the data that's 'Unknown'.

### **(c)** Are there outliers to any of the data items? What could explain those outliers? Should those records be changed or deleted?
Since my dataframe is really big, instead of searching through the numerical parts of the data table, potentially not seeing all of the outliers, I believe that it is best for me to go ahead and use the **max()** and **min()** commands in order to see if the extrema makes the most sense.

**'Year Published' Column**

Expected Minimum: `1974`

In [52]:
# Check for the lowest year value.
cleaning_df['Year Published'].min()

np.float64(1974.0)

Expected Maximum: `2024` or less.

In [53]:
# Check for the highest year value.
cleaning_df['Year Published'].max()

np.float64(2022.0)

**'Min Players' Column**

Expected Minimum: `0`or higher

At first, I thought it should be `1` or higher, but after looking more into some of the games online, some games could be autonomous where no player input is necessary. This is considered an example of a zero player game. Some board games that may be a zero player game have options where players could join in and make changes to the story lore or movement in a board game. It's very interesting and certainly something that I've never thought existed before!

In [54]:
# Check for the lowest min player value.
cleaning_df['Min Players'].min()

np.int64(0)

Expected Maximum: A reasonable amount.

In [55]:
# Check for the highest min player value.
cleaning_df['Min Players'].max()

np.int64(10)

Since 10 players isn't that much, I would say that this is a reasonable amount to be a minimum player count for a board game.

**'Max Players' Column**

Expected Minimum: `0`or higher

This would be an example of a game that is completely autonomous (see the 'Min Players' minimum check for more information about autonomous board games, which is honestly a very interesting concept).

In [56]:
# Check for the lowest max player value.
cleaning_df['Max Players'].min()

np.int64(0)

Expected Maximum: A reasonable amount.

In [57]:
# Check for the highest max player value.
cleaning_df['Max Players'].max()

np.int64(999)

At first, I didn't believe that this was a good amount. It doesn't make sense.. so I went ahead and pulled up the list of games that had a maximum player count of 999.

In [58]:
cleaning_df[cleaning_df['Max Players'] == 999]

Name  Year Published  \
7025   Start Player: A Kinda Collectible Card Game          2006.0   
8516       I Don't Know, What Do You Want to Play?          2007.0   
10813                           Scrimish Card Game          2015.0   

       Min Players  Max Players  Play Time  Min Age  Users Rated  \
7025             2          999          1        6          178   
8516             2          999          5        8           87   
10813            2          999        100        8          254   

       Rating Average  Complexity Average Owned Users Mechanics  \
7025             6.49                1.00       286.0   Unknown   
8516             6.76                1.06       117.0    Voting   
10813            5.90                1.13       898.0    Memory   

                         Domains  
7025   Family Games, Party Games  
8516                     Unknown  
10813                    Unknown

After doing more research into these games, these are board games (mainly trading card games) where over time you have to trade with other people. This would make sense, as people can infinitely trade with other people, which can explain the large max player count. In my opinion, I believe that this should stay and not be removed. **However, when I do my data analysis, I will omit these values since they can be infinite and mess with our data collection.**

**'Play Time' Column**

Expected Minimum: A reasonable amount.

In [59]:
# Check for the lowest play time value.
cleaning_df['Play Time'].min()

np.int64(0)

This is not possible as there can not be a game that has 0 playtime. I will remove these rows from the data table as the row is likely not accurate and would likely mess with our analysis.

In [60]:
# Taking a quick look at the games that have a play time of 0.
cleaning_df[cleaning_df['Play Time'] == 0]

Name  Year Published  \
5178                                             Top Ten          2020.0   
6466                                     Flesh and Blood          2019.0   
7361   Star Fleet Battles Silver Anniversary Master R...          2004.0   
7381     The Battles of Mollwitz 1741 and Chotusitz 1742          2017.0   
7588                                       Footy Manager          2011.0   
...                                                  ...             ...   
20010                       Legends of the Hidden Temple          2017.0   
20039                              WTF: (What the Fish!)          2017.0   
20068                                              W.W.B          2011.0   
20073                                       Funny or Die          2013.0   
20166                                  24 DVD Board Game          2006.0   

       Min Players  Max Players  Play Time  Min Age  Users Rated  \
5178             4            9          0       14          127   
6466             2            2          0        0           79   
7361             0            0          0       12           52   
7381             2            2          0        0           55   
7588             1            8          0        6           82   
...            ...          ...        ...      ...          ...   
20010            4           12          0       12           91   
20039            2            6          0        8           99   
20068            2            2          0        0           41   
20073            3            6          0       13           88   
20166            2            2          0       12           73   

       Rating Average  Complexity Average Owned Users  \
5178             7.90                1.00       314.0   
6466             8.68                3.80       172.0   
7361             8.59                4.86       117.0   
7381             8.51                4.50       301.0   
7588             8.06                2.00       104.0   
...               ...                 ...         ...   
20010            3.82                1.50       629.0   
20039            3.76                1.00       381.0   
20068            1.43                3.00         3.0   
20073            3.56                1.14       359.0   
20166            2.24                1.00       202.0   

                                               Mechanics  \
5178                                    Cooperative Game   
6466                                     Hand Management   
7361   Dice Rolling, Hexagon Grid, Simultaneous Actio...   
7381              Dice Rolling, Hexagon Grid, Simulation   
7588          Dice Rolling, Paper-and-Pencil, Simulation   
...                                                  ...   
20010                     Cooperative Game, Dice Rolling   
20039                         Hand Management, Take That   
20068                 Dice Rolling, Roll / Spin and Move   
20073                                             Voting   
20166                                            Unknown   

                              Domains  
5178                          Unknown  
6466                          Unknown  
7361                         Wargames  
7381                         Wargames  
7588                   Thematic Games  
...                               ...  
20010                         Unknown  
20039                         Unknown  
20068  Abstract Games, Thematic Games  
20073                         Unknown  
20166                         Unknown  

[511 rows x 12 columns]

In [61]:
# Remove all rows that have a play time value of 0.
cleaning_df = cleaning_df[cleaning_df['Play Time'] != 0]

cleaning_df

Name  Year Published  Min Players  \
0                                 Gloomhaven          2017.0            1   
1                  Pandemic Legacy: Season 1          2015.0            2   
2                          Brass: Birmingham          2018.0            2   
3                          Terraforming Mars          2016.0            1   
4          Twilight Imperium: Fourth Edition          2017.0            3   
...                                      ...             ...          ...   
20316  The Worst-Case Scenario Survival Game          2001.0            2   
20325             The Oregon Trail Card Game          2016.0            2   
20326                    Battle of the Sexes          1997.0            2   
20328                   Hungry Hungry Hippos          1978.0            2   
20334                                    LCR          1983.0            3   

       Max Players  Play Time  Min Age  Users Rated  Rating Average  \
0                4        120       14        42055            8.79   
1                4         60       13        41643            8.61   
2                4        120       14        19217            8.66   
3                5        120       12        64864            8.43   
4                6        480       14        13468            8.70   
...            ...        ...      ...          ...             ...   
20316           10         40        8          588            3.32   
20325            6         45       12         2654            4.27   
20326            8         45       12         1090            3.48   
20328            4         10        4         2361            4.12   
20334           12         20        5         1835            3.42   

       Complexity Average Owned Users  \
0                    3.86     68323.0   
1                    2.84     65294.0   
2                    3.91     28785.0   
3                    3.24     87099.0   
4                    4.22     16831.0   
...                   ...         ...   
20316                1.20      1148.0   
20325                1.18      7183.0   
20326                1.08      1987.0   
20328                1.05      2568.0   
20334                1.05      3441.0   

                                               Mechanics  \
0      Action Queue, Action Retrieval, Campaign / Bat...   
1      Action Points, Cooperative Game, Hand Manageme...   
2      Hand Management, Income, Loans, Market, Networ...   
3      Card Drafting, Drafting, End Game Bonuses, Han...   
4      Action Drafting, Area Majority / Influence, Ar...   
...                                                  ...   
20316                                       Dice Rolling   
20325  Communication Limits, Connections, Cooperative...   
20326                                    Team-Based Game   
20328                                            Unknown   
20334                                       Dice Rolling   

                              Domains  
0      Strategy Games, Thematic Games  
1      Strategy Games, Thematic Games  
2                      Strategy Games  
3                      Strategy Games  
4      Strategy Games, Thematic Games  
...                               ...  
20316                     Party Games  
20325                    Family Games  
20326                     Party Games  
20328                Children's Games  
20334   Children's Games, Party Games  

[18983 rows x 12 columns]

As a final test, I have gone ahead and checked what the new lowest play time is.

In [62]:
# Check for the lowest play time value.
cleaning_df['Play Time'].min()

np.int64(1)

While a minute does seem low, do note all of the games that could be a minute long - like charades. While the base game itself seems long, each round is short, which is what BGG (the dataset provider bases their dataset off on). I don't blame them, as this is the smartest choice to avoid providing inaccurate results.

Expected Maximum: A reasonable amount.

In [63]:
# Check for the highest play time value.
cleaning_df['Play Time'].max()

np.int64(60000)

This is very odd, so I have decided to take a quick look at the data.

In [64]:
# Taking a quick look at the games that have a play time of 0.
cleaning_df[cleaning_df['Play Time'] == 60000]

Name  Year Published  \
13420  The Campaign for North Africa: The Desert War ...          1979.0   

       Min Players  Max Players  Play Time  Min Age  Users Rated  \
13420            8           10      60000       14          146   

       Rating Average  Complexity Average Owned Users  \
13420             6.1                4.71       385.0   

                                    Mechanics   Domains  
13420  Dice Rolling, Hexagon Grid, Simulation  Wargames

From Google, 'The Campaign for North Africa: The Desert War' was a game made to be long term, lasting over a few months. While this isn't ideal, it is a piece of data that we would still need to consider. I believe that including this data would make our analysis accurate as people did play this game over a long period of time. If you are worried about our analysis making it seem like a majority of games take a long time, there are many more entires with reasonable player counts than unreasonable (beauty of having a large dataset).

**'Users Rated' Column**

Expected Minimum: A value greater than `0`.

In [65]:
# Check for the lowest users rated value.
cleaning_df['Users Rated'].min()

np.int64(30)

Expected Maximum: A reasonable amount.

In [66]:
# Check for the highest users rated value.
cleaning_df['Users Rated'].max()

np.int64(102214)

**'Rating Average' Column**

Expected Minimum: `0` or higher.

In [67]:
# Check for the lowest average rating value.
cleaning_df['Rating Average'].min()

np.float64(1.05)

Expected Maximum: `10` or lower.

In [68]:
# Check for the highest average rating value.
cleaning_df['Rating Average'].max()

np.float64(9.54)

**'Complexity Average' Column**

Expected Minimum: `0` or higher.

In [69]:
# Check for the lowest complexity rating value.
cleaning_df['Complexity Average'].min()

np.float64(0.0)

Expected Maximum: `5` or lower.

In [70]:
# Check for the highest average rating value.
cleaning_df['Complexity Average'].max()

np.float64(5.0)

**'Owned Users' Column**

Expected Minimum: A value greater than `0`.

In [71]:
# Check for the lowest owned users value.
cleaning_df['Owned Users'][cleaning_df['Owned Users'] != 'Unknown'].min()

3.0

Expected Maximum: A reasonable amount.

In [72]:
# Check for the highest owned users value.
cleaning_df['Owned Users'][cleaning_df['Owned Users'] != 'Unknown'].max()

155312.0

All of the tests above seem good, with some slight issues; however, I have addressed those issues by either explaining my standpoint and how I will handle all of the issues or handling them directly (such as when I needed to remove values that couldn't exist). If you don't think that the way I handled a situation is correct in terms of the outliers that I have, please do let me know! I am very interested in hearing what you would have done that could possibly make the results of the analysis a lot more accurate.

### **(d)** Are the column names not in a format that makes it easy to refer to?
Unfortunately, the column names that the dataset was provided with doesn't make it easy to refer to. I will rename all of the columns and rename them to all be lowercase and go ahead and replace spaces with a `_`. The good thing is that all of the column names are very short and I don't need to trim down the names themselves; I just need to go ahead and change the way they are formatted.

In [73]:
# Just for me to quickly be able to see the names of the columns without scrolling back up.
cleaning_df.columns

Index(['Name', 'Year Published', 'Min Players', 'Max Players', 'Play Time',
       'Min Age', 'Users Rated', 'Rating Average', 'Complexity Average',
       'Owned Users', 'Mechanics', 'Domains'],
      dtype='object')

In [74]:
cleaning_df.rename(columns={
    "Name": "name",
    "Year Published": "year_published",
    "Min Players": "min_players",
    "Max Players": "max_players",
    "Play Time": "play_time",
    "Min Age": "min_age",
    "Users Rated": "users_rated",
    "Rating Average": "rating_average",
    "Complexity Average": "complexity_average",
    "Owned Users": "owned_users",
    "Mechanics": "mechanics",
    "Domains": "domains"
    }, inplace=True)


C:\Users\andys\AppData\Local\Temp\ipykernel_59340\2361007550.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cleaning_df.rename(columns={


In [75]:
# To see if all of the columns have successfully been renamed.
cleaning_df.columns

Index(['name', 'year_published', 'min_players', 'max_players', 'play_time',
       'min_age', 'users_rated', 'rating_average', 'complexity_average',
       'owned_users', 'mechanics', 'domains'],
      dtype='object')

All of the columns have now been renamed to be easier for us to reference.

### **(e)** Do you need to merge multiple datasets into one?
I have only used 1 dataset for my final project, so no, I do not need to merge multiple datasets into one.

### **(f)** Do you need to split up a column's data into multiple columns or simplify its values?

The 'Mechanics' and 'Domain' columns have multiple different attributes. All of the attributes, if there are more than 1, are separated by commas, which makes the data not tidy. For this reason, we would need to figure out how to make the data tidy. Before I go ahead and figure out how I could do this, I would first like to know the unique values that could appear as an attribute for each column. 

First off, I will go ahead and find the unique values that could appear as an attribute for the 'Mechanics' column.

In [76]:
# Split the 'Mechanics' column into individual values using the ", " as the key.
mechanics = cleaning_df['mechanics'].str.split(', ')

# Going through each mechanic and using the set() operator to create a set of unique attributes.
unique_mechanics = set([mechanic for sublist in mechanics for mechanic in sublist])

unique_mechanics

{'Acting',
 'Action Drafting',
 'Action Points',
 'Action Queue',
 'Action Retrieval',
 'Action Timer',
 'Action/Event',
 'Advantage Token',
 'Alliances',
 'Area Majority / Influence',
 'Area Movement',
 'Area-Impulse',
 'Auction/Bidding',
 'Auction: Dexterity',
 'Auction: Dutch',
 'Auction: Dutch Priority',
 'Auction: English',
 'Auction: Fixed Placement',
 'Auction: Once Around',
 'Auction: Sealed Bid',
 'Auction: Turn Order Until Pass',
 'Automatic Resource Growth',
 'Betting and Bluffing',
 'Bias',
 'Bingo',
 'Bribery',
 'Campaign / Battle Card Driven',
 'Card Drafting',
 'Card Play Conflict Resolution',
 'Catch the Leader',
 'Chaining',
 'Chit-Pull System',
 'Closed Economy Auction',
 'Command Cards',
 'Commodity Speculation',
 'Communication Limits',
 'Connections',
 'Constrained Bidding',
 'Contracts',
 'Cooperative Game',
 'Crayon Rail System',
 'Critical Hits and Failures',
 'Cube Tower',
 'Deck Bag and Pool Building',
 'Deck Construction',
 'Deduction',
 'Delayed Purchase',
 

There are a lot of different attributes that could be inside of the 'Mechanics' column. I will use the length function with len() to find the total amount of unique game mechanics.

In [77]:
# Find the total amount of unique mechanics.
len(unique_mechanics)

183

There are 183 total different values that could be in the column. After looking through the output quickly, I have noticed our 'Unknown' was also counted into the mix, meaning that there are 182 different game mechanics in a board game for our dataset scope.

Now, I will go ahead and find the unique values that could appear as an attribute for the 'Domains' column.

In [78]:
# Split the 'Domains' column into individual values using the ", " as the key.
domains = cleaning_df['domains'].str.split(', ')

# Going through each domain and using the set() operator to create a set of unique attributes.
unique_domains = set([domain for sublist in domains for domain in sublist])

unique_domains

{'Abstract Games',
 "Children's Games",
 'Customizable Games',
 'Family Games',
 'Party Games',
 'Strategy Games',
 'Thematic Games',
 'Unknown',
 'Wargames'}

This column has lower types of options compared to our 'Mechanics' column. I will use the length function with len() to find the total amount of unique game genres, or domains.

In [79]:
# Find the total amount of unique domains.
len(unique_domains)

9

There are 9 total different values that could be in the column. After looking through the output quickly, I have noticed our 'Unknown' was also counted into the mix, meaning that there are 8 different game genres in our dataset.

Before, I tried to clean my data with .loc() and .iloc() but it kept saying that my data was getting corrupted and every fix I did resulted in the data set being more and more fragmented. For this reason, I have decided to ask ChatGPT to help fix my code and generate code that will not corrupted my dataframe anymore. The citation is listed in the code below.

In [81]:
# Make a copy of the cleaning dataframe for safe keeping - we don't want for it to be corrupted.
cleaning_df_col = cleaning_df.copy()

# Making an empty dict of mechanics that we will begin to fill with our column data.
mechanic_data = {}

# Go through each mechanic and make a new column in our dictionary.
for mechanic in unique_mechanics:
    mechanic_data[mechanic] = cleaning_df_col['mechanics'].apply(lambda x: mechanic in x)

# Convert the dictionary to a dataframe to concat with our main dataframe.
mechanic_df = pd.DataFrame(mechanic_data).astype(int)
mechanic_df.columns = ["mechanic_" + col for col in mechanic_df.columns]

# Making an empty dict of domains that we will begin to fill with our column data.
domain_data = {}

# Go through each domain and make a new column in our dictionary.
for domain in unique_domains:
    domain_data[domain] = cleaning_df_col['domains'].apply(lambda x: domain in x)

# Convert the dictionary to a dataframe to concat with our main dataframe.
domain_df = pd.DataFrame(domain_data).astype(int)
domain_df.columns = ["domain_" + col for col in domain_df.columns]

# Concat the two new dataframes with our main dataframe - this is to avoid corrupting the dataframe by repeatedly adding columns.
cleaning_df_col = pd.concat([cleaning_df_col, mechanic_df, domain_df], axis=1)

# Drop our original columns.
cleaning_df_col = cleaning_df_col.drop(columns=['mechanics', 'domains'])

cleaning_df_col

name  year_published  min_players  \
0                                 Gloomhaven          2017.0            1   
1                  Pandemic Legacy: Season 1          2015.0            2   
2                          Brass: Birmingham          2018.0            2   
3                          Terraforming Mars          2016.0            1   
4          Twilight Imperium: Fourth Edition          2017.0            3   
...                                      ...             ...          ...   
20316  The Worst-Case Scenario Survival Game          2001.0            2   
20325             The Oregon Trail Card Game          2016.0            2   
20326                    Battle of the Sexes          1997.0            2   
20328                   Hungry Hungry Hippos          1978.0            2   
20334                                    LCR          1983.0            3   

       max_players  play_time  min_age  users_rated  rating_average  \
0                4        120       14        42055            8.79   
1                4         60       13        41643            8.61   
2                4        120       14        19217            8.66   
3                5        120       12        64864            8.43   
4                6        480       14        13468            8.70   
...            ...        ...      ...          ...             ...   
20316           10         40        8          588            3.32   
20325            6         45       12         2654            4.27   
20326            8         45       12         1090            3.48   
20328            4         10        4         2361            4.12   
20334           12         20        5         1835            3.42   

       complexity_average owned_users  ...  mechanic_Auction: Fixed Placement  \
0                    3.86     68323.0  ...                                  0   
1                    2.84     65294.0  ...                                  0   
2                    3.91     28785.0  ...                                  0   
3                    3.24     87099.0  ...                                  0   
4                    4.22     16831.0  ...                                  0   
...                   ...         ...  ...                                ...   
20316                1.20      1148.0  ...                                  0   
20325                1.18      7183.0  ...                                  0   
20326                1.08      1987.0  ...                                  0   
20328                1.05      2568.0  ...                                  0   
20334                1.05      3441.0  ...                                  0   

       domain_Strategy Games  domain_Wargames  domain_Party Games  \
0                          1                0                   0   
1                          1                0                   0   
2                          1                0                   0   
3                          1                0                   0   
4                          1                0                   0   
...                      ...              ...                 ...   
20316                      0                0                   1   
20325                      0                0                   0   
20326                      0                0                   1   
20328                      0                0                   0   
20334                      0                0                   1   

       domain_Thematic Games  domain_Children's Games  domain_Unknown  \
0                          1                        0               0   
1                          1                        0               0   
2                          0                        0               0   
3                          0                        0               0   
4                          1                        0               0   
...                   

## 3. Save your cleaned dataset as a new CSV file.
I will submit what is exported to the file below!

In [82]:
cleaning_df_col.to_csv('CleanedBGG.csv', index=False)